Sming/SmingCore/Data/CStringArray.cpp

```diff
@@ -1,55 +1,67 @@
-/****
- * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
- * Created 2015 by Skurydin Alexey
- * http://github.com/SmingHub/Sming
- * All files of the Sming Core are provided under the LGPL v3 license.
- *
- * @author: 2018 - Mikee47 <mike@sillyhouse.net>
- *
- ****/
-
-#include "CStringArray.h"
-
-bool CStringArray::add(const char* str, unsigned length)
-{
-	if(length == 0)
-		length = strlen(str);
-	if(!reserve(len + length + 1))
-		return false;
-	if(len)
-		buffer[len++] = '\0';			   // Separator between strings
-	memcpy(buffer + len, str, length + 1); // Copy final nul terminator
-	len += length;
-	++stringCount;
-	return true;
-}
-
-int CStringArray::indexOf(const char* str) const
-{
-	if(str == nullptr)
-		return -1;
-
-	unsigned index = 0;
-	for(unsigned offset = 0; offset < len; ++index) {
-		const char* s = buffer + offset;
-		if(strcasecmp(str, s) == 0)
-			return index;
-		offset += strlen(s) + 1;
-	}
-
-	return -1;
-}
-
-const char* CStringArray::getValue(unsigned index) const
-{
-	if(index < stringCount) {
-		for(unsigned offset = 0; offset < length(); --index) {
-			const char* s = buffer + offset;
-			if(index == 0)
-				return s;
-			offset += strlen(s) + 1;
-		}
-	}
-
-	return nullptr;
-}
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2015 by Skurydin Alexey
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * @author: 2018 - Mikee47 <mike@sillyhouse.net>
+ *
+ ****/
+
+#include "CStringArray.h"
+
+bool CStringArray::add(const char* str, unsigned length)
+{
+	if(length == 0)
+		length = strlen(str);
+	if(!reserve(len + length + 1))
+		return false;
+	if(len)
+		buffer[len++] = '\0';			   // Separator between strings
+	memcpy(buffer + len, str, length + 1); // Copy final nul terminator
+	len += length;
+	++stringCount;
+	return true;
+}
+
+int CStringArray::indexOf(const char* str) const
+{
+	if(str == nullptr)
+		return -1;
+
+	unsigned index = 0;
+	for(unsigned offset = 0; offset < len; ++index) {
+		const char* s = buffer + offset;
+		if(strcasecmp(str, s) == 0)
+			return index;
+		offset += strlen(s) + 1;
+	}
+
+	return -1;
+}
+
+const char* CStringArray::getValue(unsigned index) const
+{
+	if(index < count()) {
+		for(unsigned offset = 0; offset < length(); --index) {
+			const char* s = buffer + offset;
+			if(index == 0)
+				return s;
+			offset += strlen(s) + 1;
+		}
+	}
+
+	return nullptr;
+}
+
+unsigned CStringArray::count() const
```

Comment:
Can you explain here why is that change needed?

Reply:
When an instance of CStringArray is created using an assignment, e.g. CStringArray cs("first\0second"); the quantity of elements is not set. This might normally be set within the constructor but I was unable to figure out how to do that in this case so instead I ensured count() returned the correct quantity of elements in the array by deducing whether it had already been calculated and if not then one-time calculating.